README.rst
GistAPI.py
==========

This is a Python wrapper for GitHub's Gist API.

http://develop.github.com/p/gist.html

Example Usage
-------------

::

    from gistapi import Gist, Gists

    gist = Gist('d4507e882a07ac6f9f92')
    gist.description        # 'Example Gist for gist.py'
                           
    gist.created_at         # '2010/05/16 10:51:15 -0700'
    gist.public             # False
    gist.filenames          # ['exampleEmptyFile', 'exampleFile']
    gist.files              # {'exampleFile': 'Example file content.', 'exampleEmptyFile': ''} 
    gist.comments[0].body   # Test Comment.


    Gists.fetch_by_user('kennethreitz')[-1].description    # 'My .bashrc configuration'
    
    


Installation
------------

	pip install gistapi
	
Or, if you must: 

	easy_install gistapi

Roadmap
-------

* Implement Gist API methods as they are introduced
* Possibly use other hacks in the meantime
	- Gist's New Gist method? (Post to web form?)
* Possibly add command line gist interface



HISTORY.rst
History
=======

0.2.4 (2011-03-31)
------------------
* Bugfixes
* Rely on Requests
* Unittests
* http://ci.kennethreitz.com/job/gistapi-tests/

0.2.3 (2011-03-31)
------------------
* Added Gist Comments support


0.2.2 (2011-01-10)
------------------
* Updated filename handling for special filenames


0.2.1 (2010-09-01)
------------------
* StringIO support for Files
* Added Install Requirements for Pip


0.2.0 (2010-08-28)
------------------
* Support for Add, Rename, Modify, and Delete of Files!
* Proper Unicode Support
* Proper DateTime Support
* Improved Caching
* Urllib2 Support
* PEP 008/257 Compliance
* Secret `.pibb` URL Attribute for gists :)
* Misc. Refactors

0.1.4 (2010-08-18)
------------------
* Python 2.5 fix

0.1.3 (2010-06-14)
------------------
* Nose-test fix

0.1.2 (2010-05-16)
------------------
* URL Generation BugFix
* Updated versioning
* Metadata refactoring
* General Bugfixes

0.1.1 (2010-05-16)
------------------

* PyPi Compatibility Changes
* Documentation Updates
* Added DocTests

0.1.0 (2010-05-16)
------------------

* Initial Release


test_gistapi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest

import gistapi
from gistapi import Gist, Gists


class RequestsTestSuite(unittest.TestCase):
    """Requests test cases."""

    def setUp(self):
        pass

    def tearDown(self):
        """Teardown."""
        pass

    def test_repo_fetch(self):
        r1 = Gist('d4507e882a07ac6f9f92').repo
        r2 = u'd4507e882a07ac6f9f92'

        self.assertEqual(r1, r2)

    def test_owner_fetch(self):
        r1 = Gist('d4507e882a07ac6f9f92').owner
        r2 = u'kennethreitz'

        self.assertEqual(r1, r2)

    def test_created_at_fetch(self):
        r1 = Gist('d4507e882a07ac6f9f92').created_at.isoformat()
        r2 = '2010-05-16T10:51:15'

        self.assertEqual(r1, r2)

    def test_public_fetch(self):
        r1 = Gist('d4507e882a07ac6f9f92').public
        r2 = False

        self.assertEqual(r1, r2)

    def test_fetch_filesnames(self):
        r1 = Gist('d4507e882a07ac6f9f92').filenames
        r2 = ['exampleEmptyFile', 'exampleFile']

        self.assertEqual(r1, r2)

    def test_gist_search(self):
        r1 = Gists.fetch_by_user('kennethreitz')[-1].description
        r2 = u'My .bashrc configuration'

        self.assertEqual(r1, r2)

    def test_gist_comments(self):
        r1 = Gist(885658).comments[0].body
        r2 = u'Great stuff.'

        self.assertEqual(r1, r2)
    


if __name__ == '__main__':
    unittest.main()


gistapi/tests/test_gist.py
from gistapi import *

for gist in Gists.fetch_by_user('defunkt'):
    if gist.description == 'A list of Gist clients.':
        print gist.id
        print gist.repo
        print gist.embed_url


gistapi/tests/test_examples.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import doctest
import gistapi

# run doctests
doctest.testmod(gistapi.core)


gist/gist.py
#!/usr/bin/env python
# encoding: utf-8

"""
Gist.py -- A Python wrapper for the Gist API
(c) 2010 Kenneth Reitz. PSFL License.

Example usage:

>>> Gist('d4507e882a07ac6f9f92').repo
'd4507e882a07ac6f9f92'

>>> Gist('d4507e882a07ac6f9f92').description
'Example Gist for gist.py'

>>> Gist('d4507e882a07ac6f9f92').created_at
'2010/05/16 10:51:15 -0700'

>>> Gist('d4507e882a07ac6f9f92').public
False

>>> Gist('d4507e882a07ac6f9f92').filenames
['exampleEmptyFile', 'exampleFile']

>>> Gist('d4507e882a07ac6f9f92').files
{'exampleFile': 'Example file content.', 'exampleEmptyFile': ''}

>>> Gists.fetch_by_user('kennethreitz')[-1].description
'My .bashrc configuration'
"""

import urllib

try: import simplejson as json
except ImportError: import json

__author__ = 'Kenneth Reitz'
__version__ = ('0', '6', '1')
__license__ = 'MIT'


class Gist(object):
	"""Gist Object"""
	
	def __init__(self, id=None, json=None):
		self.id = id; self._json = json
		self.url = 'http://github.com/{0}'.format(id)
		self.embed_url = 'http://github.com/{0}.js'.format(id)
		self.json_url = 'http://github.com/{0}.json'.format(id)

		# Map given repo id to gist id if none exists
		if self._json: self.id = json['repo']
	
	def __getattribute__(self, name):
		"""Gets attributes, but only if needed"""

		# Only make external API calls if needed
		if name in ['description', 'created_at', 'public', 'files', 'filenames', 'repo']: 
			if not hasattr(self, '_meta'):
				self._meta = self._get_meta()
			
		return object.__getattribute__(self, name)

	def _get_meta(self):
		"""Fetches Gist metadata"""
		
		# Use json data provided if available
		if self._json:
			_meta = self._json 
			setattr(self, 'id', _meta['repo'])
		else:
			# Fetch Gist metadata
			_meta_url = 'http://gist.github.com/api/v1/json/{0}'.format(self.id)
			_meta = json.load(urllib.urlopen(_meta_url))['gists'][0]
		
		for key, value in _meta.iteritems(): 
			
			if key == 'files': 
				# Remap file key from API
				setattr(self, 'filenames', value)
			else:
				# Attach properties to object
				setattr(self, key, value)
			
		return _meta
	
	@property
	def files(self):
		"""Fetches a gists files and stores them in the 'files' property"""
		files = {}
		
		for fn in self.filenames:
			# Grab file contents
			file_url = 'http://gist.github.com/raw/{0}/{1}'.format(self.id, fn)
			files[fn] = (urllib.urlopen(file_url).read())
			
		return files
	

class Gists(object):
	"""Gist API wrapper"""
	def __init__(self, username=None, token=None):
		# Token-based Authentication is unnecesary at this point, gist api functionality is still in alpha
		self._username = username; self._token = token

	@staticmethod
	def fetch_by_user(name):
		"""Returns a set of public Gist objects owned by the given GitHub username"""

		_url = 'http://gist.github.com/api/v1/json/gists/{0}'.format(name)
		
		# Return a list of Gist objects 
		return [Gist(json=g) for g in json.load(urllib.urlopen(_url))['gists']]


if __name__ == '__main__':
	import doctest
	doctest.testmod()
	

gistapi/gistapi.py
### !/usr/bin/env python
# encoding: utf-8

"""
GistAPI.py -- A Python wrapper for the Gist API
(c) 2010 Kenneth Reitz. MIT License.

Example usage:

>>> Gist('d4507e882a07ac6f9f92').repo
'd4507e882a07ac6f9f92'

>>> Gist('d4507e882a07ac6f9f92').owner
'kennethreitz'

>>> Gist('d4507e882a07ac6f9f92').description
'Example Gist for gist.py'

>>> Gist('d4507e882a07ac6f9f92').created_at
'2010/05/16 10:51:15 -0700'

>>> Gist('d4507e882a07ac6f9f92').public
False

>>> Gist('d4507e882a07ac6f9f92').filenames
['exampleEmptyFile', 'exampleFile']

>>> Gist('d4507e882a07ac6f9f92').files
{'exampleFile': 'Example file content.', uexampleEmptyFile': ''}

>>> Gists.fetch_by_user('kennethreitz')[-1].description
'My .bashrc configuration'
"""

import os.path
import urllib
import urllib2

try:
    import simplejson as json
except ImportError:
    import json

__all__ = ['Gist', 'Gists']

# Set your own credentials
USERNAME = 'YOUR GITHUB USERNAME'
TOKEN = 'YOUR GITHUB TOKEN'

GIST_BASE = 'http://gist.github.com/%s'
GIST_JSON = GIST_BASE % 'api/v1/json/%s'


class Gist(object):
    """Gist Object"""

    def __init__(self, id=None, json=None, username=None, token=None):
        self.id = id
        self._json = json
        self._username = username
        self._token = token

        # Map given repo id to gist id if none exists
        if self._json:
            self.id = json['repo']

        self.url = url = GIST_BASE % self.id
        self.embed_url = url + '.js'
        self.json_url = url + '.json'
        self.post_url = GIST_BASE % 'gists/%s' % self.id

    def __getattribute__(self, name):
        """Get attributes, but only if needed."""

        # Only make external API calls if needed
        if name in ('owner', 'description', 'created_at', 'public',
                    'files', 'filenames', 'repo'):
            if not hasattr(self, '_meta'):
                self._meta = self._get_meta()

        return object.__getattribute__(self, name)

    def _get_meta(self):
        """Fetch Gist metadata."""

        # Use json data provided if available
        if self._json:
            _meta = self._json
            setattr(self, 'id', _meta['repo'])
        else:
            # Fetch Gist metadata
            _meta_url = GIST_JSON % self.id
            _meta = json.load(urllib2.urlopen(_meta_url))['gists'][0]

        for key, value in _meta.iteritems():

            if key == 'files':
                # Remap file key from API.  This structure stores renames.
                setattr(self, 'filenames', dict((fn, fn) for fn in value))
                # This attribute stores the {new_name: original_name} mapping.
                setattr(self, '_renames', {})
            else:
                # Attach properties to object
                setattr(self, key, value)

        return _meta

    def _post(self, params, headers={}):
        """POST to the web form (internal method)."""
        request = urllib2.Request(self.post_url,
                                  urllib.urlencode(params),
                                  headers)
        try:
            response = urllib2.urlopen(request)
        except IOError, exc:
            response = exc
        return response.code, response.msg

    def reset(self):
        """Clear the local cache."""
        if hasattr(self, '_files'):
            del self._files
        if hasattr(self, '_meta'):
            del self._meta

    def auth(self, username, token):
        """Set credentials."""
        self._username = username
        self._token = token

    def add(self, name, content=''):
        """Add file to the Gist."""
        if name in self.filenames.values():
            raise KeyError('File %r already exist' % name)
        defaultname = 'gistfile%s' % len(self.filenames)
        self.files[name] = content
        self.filenames[defaultname] = name
        self._renames[name] = defaultname

    def rename(self, from_name, to_name):
        """Rename a file."""
        filenames = self.filenames.values()
        if from_name not in filenames:
            raise KeyError('File %r does not exist' % from_name)
        if to_name in filenames:
            raise KeyError('File %r already exist' % to_name)
        orig_name = self._renames.pop(from_name, from_name)
        self.files[to_name] = self.files.pop(from_name)
        self.filenames[orig_name] = to_name
        self._renames[to_name] = orig_name

    def delete(self, name):
        """Delete a file."""
        orig_name = self._renames.pop(name, name)
        del self.files[name]
        del self.filenames[orig_name]

    def save(self):
        """Upload the changes to Github."""
        params = {
            '_method': 'put',
            'login': self._username or USERNAME,
            'token': self._token or TOKEN,
        }
        for orig, fn in self.filenames.items():
            ext = os.path.splitext(fn)[1] or '.txt'
            content = self.files[fn]
            params.update({
                'file_name[%s]' % orig: fn,
                'file_ext[%s]' % orig: ext,
                'file_contents[%s]' % orig: content,
            })
        code, msg = self._post(params=params)
        if code == 200:  # OK
            # If successful, clear the cache
            self.reset()
        return code, msg

    @property
    def files(self):
        """Fetch Gist's files and store them in the 'files' property."""
        try:
            return self._files
        except AttributeError:
            self._files = _files = {}

        for fn in self.filenames:
            # Grab file contents
            _file_url = GIST_BASE % 'raw/%s/%s' % (self.id, fn)
            _files[fn] = urllib2.urlopen(_file_url).read()

        return _files


class Gists(object):
    """Gist API wrapper"""

    def __init__(self, username=None, token=None):
        # Token-based Authentication is unnecessary, gist api still in alpha
        self._username = username
        self._token = token

    @staticmethod
    def fetch_by_user(name):
        """Return a list of public Gist objects owned by
        the given GitHub username."""

        _url = GIST_JSON % 'gists/%s' % name

        # Return a list of Gist objects
        return [Gist(json=g)
                for g in json.load(urllib2.urlopen(_url))['gists']]


if __name__ == '__main__':
    import doctest
    print('hello')
    a = 'bob'

    doctest.testmod()


gistapi/core.py
# -*- coding: utf-8 -*-

"""
GistAPI.py -- A Python wrapper for GitHub's Gist API
(c) 2011 Kenneth Reitz. MIT License.

Example usage:

>>> Gist('d4507e882a07ac6f9f92').repo
u'd4507e882a07ac6f9f92'

>>> Gist('d4507e882a07ac6f9f92').owner
u'kennethreitz'

>>> Gist('d4507e882a07ac6f9f92').description
u'Example Gist for gist.py'

>>> Gist('d4507e882a07ac6f9f92').created_at.isoformat()
'2010-05-16T10:51:15-07:00'

>>> Gist('d4507e882a07ac6f9f92').public
False

>>> Gist('d4507e882a07ac6f9f92').filenames
['exampleEmptyFile', 'exampleFile']

>>> Gists.fetch_by_user('kennethreitz')[-1].description
u'My .bashrc configuration'

>>> Gist(885658).comments[0].body
u'Great Stuff.'
"""


import cStringIO
import os.path

import requests
import urllib2
from datetime import datetime

try:
    import simplejson as json
except ImportError:
    import json

__all__ = ['Gist', 'Gists']

# Set your own credentials
USERNAME = 'YOUR GITHUB USERNAME'
TOKEN = 'YOUR GITHUB TOKEN'

GIST_BASE = 'http://gist.github.com/%s'
GIST_JSON = GIST_BASE % 'api/v1/json/%s'


class Gist(object):
    """Gist Object"""

    def __init__(self, id=None, json=None, username=None, token=None):
        self.id = id
        self._json = json
        self._username = username
        self._token = token

        # Map given repo id to gist id if none exists
        if self._json:
            self.id = json['repo']

        self.url = url = GIST_BASE % self.id
        self.embed_url = url + '.js'
        self.epic_embed_url = url + '.pibb'
        self.json_url = url + '.json'
        self.post_url = GIST_BASE % 'gists/%s' % self.id
        self.comments  = []
    
    def __repr__(self):
        return '<gist %s>' % self.id

    def __getattribute__(self, name):
        """Get attributes, but only if needed."""

        # Only make external API calls if needed
        if name in ('owner', 'description', 'created_at', 'public',
                    'files', 'filenames', 'repo', 'comments'):
            if not hasattr(self, '_meta'):
                self._meta = self._get_meta()

        return object.__getattribute__(self, name)

    def _get_meta(self):
        """Fetch Gist metadata."""

        # Use json data provided if available
        if self._json:
            _meta = self._json
            setattr(self, 'id', _meta['repo'])
        else:
            # Fetch Gist metadata
            _meta_url = GIST_JSON % self.id
            _meta = json.loads(requests.get(_meta_url).content)['gists'][0]

        for key, value in _meta.iteritems():

            if key == 'files':
                # Remap file key from API.
                setattr(self, 'filenames', value)
                # Store the {current_name: original_name} mapping.
                setattr(self, '_renames', dict((fn, fn) for fn in value))
            elif key == 'public':
                # Attach booleans
                setattr(self, key, value)
            elif key == 'created_at':
                # Attach datetime
                setattr(self, 'created_at', datetime.strptime(value[:-6], '%Y/%m/%d %H:%M:%S'))
                
            elif key == 'comments':
                _comments = []
                for comment in value:
                    c = GistComment().from_api(comment)
                    _comments.append(c)
                    setattr(self, 'comments', _comments)
            else:
                # Attach properties to object
                setattr(self, key, unicode(value))

        return _meta

    def _post(self, params, headers={}):
        """POST to the web form (internal method)."""
        r = requests.post(self.post_url, params, headers=headers)

        return r.status_code, r.content

    def reset(self):
        """Clear the local cache."""
        if hasattr(self, '_files'):
            del self._files
        if hasattr(self, '_meta'):
            del self._meta

    def auth(self, username, token):
        """Set credentials."""
        self._username = username
        self._token = token

    def rename(self, from_name, to_name):
        """Rename a file."""
        if from_name not in self.files:
            raise KeyError('File %r does not exist' % from_name)
        if to_name in self.files:
            raise KeyError('File %r already exist' % to_name)
        self.files[to_name] = self.files.pop(from_name)
        try:
            self._renames[to_name] = self._renames.pop(from_name)
        except KeyError:
            # New file
            pass

    def save(self):
        """Upload the changes to Github."""
        params = {
            '_method': 'put',
            'login': self._username or USERNAME,
            'token': self._token or TOKEN,
        }
        names_map = self._renames
        original_names = names_map.values()
        index = len(original_names)
        for fn, content in self.files.items():
            ext = os.path.splitext(fn)[1] or '.txt'
            try:
                orig = names_map.pop(fn)
            except KeyError:
                # Find a unique filename
                while True:
                    orig = 'gistfile%s' % index
                    index += 1
                    if orig not in original_names:
                        break
            params.update({
                'file_name[%s]' % orig: fn,
                'file_ext[%s]' % orig: ext,
                'file_contents[%s]' % orig: content,
            })
        code, msg = self._post(params=params)
        if code == 200:  # OK
            # If successful, clear the cache
            self.reset()
        return code, msg

    @property
    def files(self):
        """Fetch Gist's files and store them in the 'files' property."""
        try:
            return self._files
        except AttributeError:
            self._files = _files = {}

        for fn in self._meta['files']:
            # Grab file contents
            _file_url = GIST_BASE % 'raw/%s/%s' % (self.id, urllib2.quote(fn))
            _files[fn] = cStringIO.StringIO()
            _files[fn].write(requests.get(_file_url).content)

        return _files


class Gists(object):
    """Gist API wrapper"""

    def __init__(self, username=None, token=None):
        # Token-based Authentication is unnecessary, gist api still in alpha
        self._username = username
        self._token = token

    @staticmethod
    def fetch_by_user(name):
        """Return a list of public Gist objects owned by
        the given GitHub username."""

        _url = GIST_JSON % 'gists/%s' % name

        # Return a list of Gist objects
        return [Gist(json=g)
                for g in json.loads(requests.get(_url).content)['gists']]


class GistComment(object):
    """Gist comments."""
    
    def __init__(self): 
        self.body = None
        self.created_at = None
        self.gravatar_id = None
        self.id = None
        self.updated_at = None
        self.user = None

    def __repr__(self):
        return '<gist-comment %s>' % self.id

    @staticmethod
    def from_api(jsondict):
        """Returns new instance of GistComment containing given api dict."""
        comment = GistComment()
        
        comment.body = jsondict.get('body', None)
        comment.created_at = datetime.strptime(jsondict.get('created_at')[:-6], '%Y/%m/%d %H:%M:%S')
        comment.gravatar_id = jsondict.get('gravatar_id', None)
        comment.id = jsondict.get('id', None)
        comment.updated_at = datetime.strptime(jsondict.get('updated_at')[:-6], '%Y/%m/%d %H:%M:%S')
        comment.user = jsondict.get('user', None)
        
        return comment
        
